src/plugins/plugin.gridLines.js

```diff
@@ -0,0 +1,308 @@
+				y2 = chartArea.bottom;
+			} else {
+				var yLineValue = scale.getPixelForTick(index);
+
+				x1 = chartArea.left;
+				x2 = chartArea.right;
+				y1 = y2 = yLineValue;
+			}
+
+			context.lineWidth = lineWidth;
+			context.strokeStyle = lineColor;
+			if (context.setLineDash) {
+				context.setLineDash(borderDash);
+				context.lineDashOffset = borderDashOffset;
+			}
+
+			// Draw current gridLine
+			drawLine(context, x1, x2, y1, y2);
+		}
+	}
+
+	function drawBorders(context, bordersToDraw, undefinedBorderOptions) {
+		// Draws all the borders. Skips undefined orders which would be overlapped by a defined border
+		helpers.each(bordersToDraw, function(borderToDraw) {
+			// When the border is undefined, check if there isn't a defined border on the same place, if there is one dont draw this border
+			if (!borderToDraw.undefinedBorder || bordersToDraw.findIndex(bordersOverlap, borderToDraw) === -1) {
+				context.save();
+
+				var _undefinedBorderOptions = borderToDraw.isHorizontal ? undefinedBorderOptions.horizontal : undefinedBorderOptions.vertical;
+
+				// Use given properties if border is defined or they were explictly set(e.g. when the undefined border is also a zeroLine)
+				// Otherwise use default properties for undefined borders
+				var useGivenOptions = !borderToDraw.undefinedBorder || borderToDraw.forceUseStyle;
+
+				context.lineWidth = useGivenOptions ? borderToDraw.lineWidth : _undefinedBorderOptions.lineWidth;
+				context.strokeStyle = useGivenOptions ? borderToDraw.lineColor : _undefinedBorderOptions.lineColor;
+				if (context.setLineDash) {
+					context.setLineDash(useGivenOptions ? borderToDraw.borderDash : _undefinedBorderOptions.borderDash);
+					context.lineDashOffset = useGivenOptions ? borderToDraw.borderDashOffset : _undefinedBorderOptions.borderDashOffset;
+				}
+
+				// Draw the border
+				drawLine(context, borderToDraw.x1, borderToDraw.x2, borderToDraw.y1, borderToDraw.y2);
+
+				// If there are no defined borders overlapping this undefined border, mark it
+				// as defined to prevent other undefined borders to be drawn over it
+				if (borderToDraw.undefinedBorder) {
+					borderToDraw.undefinedBorder = false;
+				}
+			}
+		});
+	}
+
+	function setUndefinedBorderOptionsToDefault(undefinedBorderOptions) {
+		var gridLineDefaults = Chart.defaults.scale.gridLines;
+
+		if (undefinedBorderOptions.horizontal === undefined) {
+			undefinedBorderOptions.horizontal = {
+				lineWidth: gridLineDefaults.lineWidth,
+				lineColor: gridLineDefaults.color,
+				borderDash: gridLineDefaults.borderDash,
+				borderDashOffset: gridLineDefaults.borderDashOffset
+			};
+		}
+
+		if (undefinedBorderOptions.vertical === undefined) {
+			undefinedBorderOptions.vertical = {
+				lineWidth: gridLineDefaults.lineWidth,
+				lineColor: gridLineDefaults.color,
+				borderDash: gridLineDefaults.borderDash,
+				borderDashOffset: gridLineDefaults.borderDashOffset
+			};
+		}
+	}
+
+	return {
+		id: 'gridLines',
```

Comment:
I still need to fully review that PR but a first feedback according the [plugin guidelines](/chartjs/Chart.js/blob/master/docs/developers/plugins.md#plugin-id): the id should be lowercase (`id: 'gridlines'`). Can you also rename the file to: `src/plugins/plugin.gridlines.js`?